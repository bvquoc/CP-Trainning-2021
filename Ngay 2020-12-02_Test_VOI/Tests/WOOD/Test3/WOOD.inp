953094027 2 7
1052662 0 1052662 1
18575012 0 18575012 1
0 1 361074435 1
542357498 0 542357498 2
566253429 0 566253429 2
361074435 0 361074435 2
9986990 0 9986990 1
