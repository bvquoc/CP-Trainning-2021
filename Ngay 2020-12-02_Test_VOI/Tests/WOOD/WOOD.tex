\begin{problem}{WOOD}{WOOD.inp}{WOOD.out}{1 giây}{512M}
\small{
\addcontentsline{toc}{subsection}{WOOD}

Hùng vừa cắt một tấm gỗ mỏng hình chữ nhật thành nhiều miếng hình chữ nhật. Máy cưa cậu dùng mỗi lần sẽ nhận vào một miếng gỗ hình chữ nhật và cắt làm hai hình chữ nhật với tổng diện tích không đổi. Vị trí cắt có thể điều chỉnh được, chi phí cắt phụ thuộc vào trọng lượng miếng gỗ, do miếng gỗ ban đầu có độ dày đồng đều nên chi phí cắt có thể xem là diện tích miếng gỗ đó.\\
\begin{center}
	\ig{WOOD/WOOD.png}
\end{center}
Miếng gỗ ban đầu có kích thước $n \times m$. Hùng đã dùng bút vẽ $n-1$ đường kẻ ngang và $m-1$ đường kẻ dọc, kết hợp với các đường biên chia hình chữ nhật thành một lưới tọa độ. Các điểm được đánh tọa độ từ $(0,0)$ đến $(n,m)$ từ trên xuống dưới, từ trái qua phải. Các lát cắt đều song song với biên và được đánh dấu bởi tọa độ các điểm này. Mỗi lát cắt là một đường liên tục và cắt dứt điểm một miếng nào đó làm hai phần. Miếng gỗ ban đầu đã được cắt xong, Hùng cần báo cáo lại quá trình cắt. Cậu nhớ chính xác tọa độ tất cả các lát cắt mà mình thực hiện, tuy nhiên không nhớ chính xác thứ tự đã cắt. Hùng đưa cho bạn danh sách các lát cắt này và nhờ bạn sắp xếp lại thứ tự cho hợp lý. Cậu cũng biết rằng có thể có nhiều thứ tự khác nhau, và quả quyết rằng trong số các thứ tự đúng đó, mình đã cắt theo một thứ có tổng chi phí cắt là nhỏ nhất.\\
\textbf{Yêu cầu:} Hãy giúp Hùng tìm ra một thứ tự cắt (tức hoán vị các lát cắt theo trí nhớ của Hùng) hợp lý và có tổng chi phí nhỏ nhất, hoặc thông báo là cậu đã nhớ nhầm.

\InputFile
\begin{itemize}
	\item Dòng đầu chứa ba số nguyên dương là kích thước ban đầu và số lát cắt: $n$ $m$ $k$ ($1 \leq n,m \leq 10^9$, $1 \leq k \leq 10^5$)
	\item $k$ dòng tiếp theo, mỗi dòng ghi một lát cắt: $u$ $v$ $x$ $y$ ($0 \leq u \leq x \leq n, 0 \leq v \leq y \leq m$) có nghĩa lát cắt này kéo dài từ điểm $(u,v)$ đến điểm $(x,y)$. Dữ liệu đảm bảo lát cắt song song với biên, không trùng lên biên của hình chữ nhật đang cắt và có độ dài khác 0.
\end{itemize}

\OutputFile
Nếu không có thứ tự nào hợp lý, in ra -1. Ngược lại in ra trên hai dòng:
\begin{itemize}
	\item Dòng đầu chứa chi phí cắt nhỏ nhất tìm được
	\item Dòng tiếp theo chứa $k$ số là một hoán vị của $\{1, 2, \ldots, k\}$ cho biết thứ tự các lát cắt tìm được.
\end{itemize}
Dữ liệu đảm bảo chi phí nhỏ nhất nếu có sẽ không vượt quá $10^{18}$

\Examples
\begin{example}%
\exmp{
5 10 11
3 5 5 5
3 0 3 10
0 5 3 5
1 5 1 10
2 5 2 10
4 5 4 10
3 1 5 1
3 2 5 2
3 3 5 3
3 4 5 4
1 8 2 8
}{
164
2 1 3 4 5 6 8 7 9 11 10
}%
\end{example}
%\Explanation
\Scoring
Với mỗi test, nếu chỉ tính đúng chi phí cắt sẽ được 50\% điểm của test đó, để nhận được 50\% điểm này thì dòng thứ hai bạn cần để trống.
\begin{itemize}
	\item Subtask 1 (20\%): $k \leq 8$
	\item Subtask 2 (20\%): $n = 1$
	\item Subtask 3 (20\%): Thứ tự Hùng nhớ là một thứ tự hợp lý và tối ưu
	\item Subtask 4 (20\%): Thứ tự Hùng nhớ là một thứ tự hợp lý
	\item Subtask 5 (20\%): Ràng buộc gốc
\end{itemize}

\begin{comment}
\Hint

\end{comment}
}
\end{problem}
