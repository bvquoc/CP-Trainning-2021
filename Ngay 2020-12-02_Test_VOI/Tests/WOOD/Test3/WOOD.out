4560838207
5 6 4 3 2 7 1 